File: data_structures/stacks/stack.cpp
#include <iostream>
#include <stdexcept>
#include <vector>

// Generic Stack using vector
template <typename T> class Stack {
private:
  std::vector<T> stack;
  int limit;

public:
  Stack(int limit = 10) : limit(limit) {}

  void push(T data) {
    if (stack.size() >= limit) {
      throw std::overflow_error("Stack Overflow");
    }
    stack.push_back(data);
  }

  T pop() {
    if (stack.empty()) {
      throw std::underflow_error("Stack Underflow");
    }
    T data = stack.back();
    stack.pop_back();
    return data;
  }

  T peek() const {
    if (stack.empty()) {
      throw std::underflow_error("Stack Underflow");
    }
    return stack.back();
  }

  bool isEmpty() const { return stack.empty(); }

  bool isFull() const { return stack.size() == limit; }

  int size() const { return stack.size(); }
};

int main() {
  Stack<int> s(5);
  s.push(1);
  s.push(2);
  s.push(3);

  std::cout << "Top: " << s.peek() << std::endl;
  std::cout << "Popped: " << s.pop() << std::endl;
  std::cout << "Size: " << s.size() << std::endl;

  return 0;
}

File: data_structures/stacks/stack.rs
#[derive(Debug)]
pub struct Stack<T> {
    stack: Vec<T>,
    limit: usize,
}

impl<T> Stack<T> {
    pub fn new(limit: usize) -> Self {
        Stack {
            stack: Vec::new(),
            limit,
        }
    }

    pub fn push(&mut self, data: T) -> Result<(), &'static str> {
        if self.stack.len() >= self.limit {
            return Err("Stack Overflow");
        }
        self.stack.push(data);
        Ok(())
    }

    pub fn pop(&mut self) -> Option<T> {
        self.stack.pop()
    }

    pub fn peek(&self) -> Option<&T> {
        self.stack.last()
    }

    pub fn is_empty(&self) -> bool {
        self.stack.is_empty()
    }

    pub fn is_full(&self) -> bool {
        self.stack.len() == self.limit
    }

    pub fn size(&self) -> usize {
        self.stack.len()
    }
}

fn main() {
    let mut s = Stack::new(5);
    s.push(1).unwrap();
    s.push(2).unwrap();
    s.push(3).unwrap();

    println!("Top: {:?}", s.peek());
    println!("Popped: {:?}", s.pop());
    println!("Size: {}", s.size());
}
